File: 프로그래머스/lv2/81302. 거리두기 확인하기/거리두기 확인하기.js
const dx = [1, 0, -1, 0];
const dy = [0, 1, 0, -1];
function bfs(x, y, place, chkb, n, m) {
    chkb[x][y] = true;
    const q = [];
    q.push([x, y])
    while (q.length) {
        const [ex, ey] = q.splice(0, 1)[0];
        for (let k = 0; k < 4; k += 1) {
            let nx = ex + dx[k];
            let ny = ey + dy[k];
            if (0<=nx && nx<n && 0<=ny && ny<m && chkb[nx][ny] == false) {
                if (place[nx][ny] == "P") {
                    return false
                }
                if (place[nx][ny] == "O" && place[ex][ey] != "O") {
                    chkb[nx][ny] = true;
                    q.push([nx, ny])
                }
            }
        }
    }

    return true
}

function solution(places) {
    var answer = [];

    var n = places.length;
    var m = places[0].length;

    for (const place of places) {
        const chkb = Array.from(Array(n), () => Array(m).fill(false));
        let discrim = true;
        for (let i = 0; i < n; i += 1) {
            for (let j = 0; j < m; j += 1) {
                if (place[i][j] == "P" && chkb[i][j] == false) {
                    if (bfs(i, j, place, chkb, n, m) == false) {
                        discrim = false;
                        break;
                    }
                }
            }
        }
        if (discrim == true) {
            answer.push(1)
        } else {
            answer.push(0)
        }
    }

    return answer;
}

console.log(solution([["POOOP", "OXXOX", "OPXPX", "OOXOX", "POXXP"], ["POOPX", "OXPXP", "PXXXO", "OXXXO", "OOOPP"], ["PXOPX", "OXOXP", "OXPOX", "OXXOP", "PXPOX"], ["OOOXX", "XOOOX", "OOOXX", "OXOOX", "OOOOO"], ["PXPXP", "XPXPX", "PXPXP", "XPXPX", "PXPXP"]]))

File: 프로그래머스/lv2/81302. 거리두기 확인하기/거리두기 확인하기.py
from collections import deque

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]

def bfs(x, y, graph, chkb):
    q = deque()
    q.append((x, y))
    chkb[x][y] = True
    while q:
        ex, ey = q.popleft()
        for k in range(4):
            nx = ex + dx[k]
            ny = ey + dy[k]
            if 0<=nx<n and 0<=ny<m and not chkb[nx][ny]:
                if graph[nx][ny] == "O" and graph[ex][ey] != "O":
                    q.append((nx, ny))
                    chkb[nx][ny] = True
                elif graph[nx][ny] == "P":
                    return False
    return True

def solution(places):
    global n, m
    answer = []
    n = len(places)
    m = len(places[0])

    for place in places:
        chkb = [[False] * m for _ in range(n)]
        discrim = True
        for i in range(n):
            for j in range(m):
                if not chkb[i][j] and place[i][j] == "P":
                    chkb[i][j] = True
                    if not bfs(i, j, place, chkb):
                        discrim = False
                        break
        if discrim:
            answer.append(1)
        else:
            answer.append(0)

    return answer